pkg/collector/scheduler/job.go

```diff
@@ -10,97 +10,187 @@ import (
+		nb = 1
+	} else {
+		nb = int(interval.Truncate(time.Second).Seconds())
+	}
+	for i := 0; i < nb; i++ {
+		bucket := &jobBucket{}
+		jq.buckets = append(jq.buckets, bucket)
+	}
+
+	// compute step for sparse scheduling
+	switch nb % 2 {
+	case 0:
+		step := nb / 2
+		switch step % 2 {
+		case 0:
+			jq.sparseStep = uint(step - 1)
+		case 1:
+			jq.sparseStep = uint(step - 2)
+		}
+	case 1:
+		jq.sparseStep = uint(nb / 2)
 	}
 
+	return jq
 }
 
 // addJob is a convenience method to add a check to a queue
 func (jq *jobQueue) addJob(c check.Check) {
 	jq.mu.Lock()
 	defer jq.mu.Unlock()
 
-	jq.jobs = append(jq.jobs, c)
+	// Checks scheduled to buckets scheduled with sparse round-robin
+	jq.buckets[jq.schedulingBucketIdx].addJob(c)
+	jq.schedulingBucketIdx = (jq.schedulingBucketIdx + jq.sparseStep) % uint(len(jq.buckets))
 }
 
 func (jq *jobQueue) removeJob(id check.ID) error {
 	jq.mu.Lock()
 	defer jq.mu.Unlock()
 
-	for i, c := range jq.jobs {
-		if c.ID() == id {
-			jq.jobs = append(jq.jobs[:i], jq.jobs[i+1:]...)
-			return nil
+	for i, bucket := range jq.buckets {
+		bucket.mu.Lock()
+		for j, c := range bucket.jobs {
+			if c.ID() == id {
+				jq.buckets[i].jobs = append(bucket.jobs[:j], bucket.jobs[j+1:]...)
+				bucket.mu.Unlock()
+				return nil
+			}
 		}
+		bucket.mu.Unlock()
 	}
 
 	return fmt.Errorf("check with id %s is not in this Job Queue", id)
 }
 
+func (jq *jobQueue) stats() map[string]interface{} {
+	jq.mu.RLock()
+	defer jq.mu.RUnlock()
+
+	nJobs := 0
+	nBuckets := 0
+	for _, bucket := range jq.buckets {
+		nJobs += bucket.size()
+		nBuckets++
+	}
+
+	return map[string]interface{}{
+		"Interval": jq.interval / time.Second,
+		"Buckets":  nBuckets,
+		"Size":     nJobs,
+	}
+}
+
 // run schedules the checks in the queue by posting them to the
 // execution pipeline.
 // Not blocking, runs in a new goroutine.
 func (jq *jobQueue) run(out chan<- check.Check) {
+
 	go func() {
-		for jq.waitForTick(out) {
+		log.Debugf("Job queue is running...")
+		for jq.process(out) {
+			// empty
 		}
 		jq.stopped <- true
 	}()
 }
 
-// waitForTicks enqueues the checks at a tick, and returns whether the queue
+// process  enqueues the checks at a tick, and returns whether the queue
 // should listen to the following tick (or stop)
-func (jq *jobQueue) waitForTick(out chan<- check.Check) bool {
+func (jq *jobQueue) process(out chan<- check.Check) bool {
+
 	select {
 	case <-jq.stop:
-		// someone asked to stop this queue
-		jq.ticker.Stop()
 		jq.health.Deregister()
 		return false
-	case <-jq.health.C:
-	case <-jq.ticker.C:
-		// normal case, (re)schedule the queue
-		jq.mu.RLock()
-		for _, check := range jq.jobs {
-			// sending to `out` is blocking, we need to constantly check that someone
-			// isn't asking to stop this queue
+	case t := <-jq.bucketTicker.C:
+		log.Debugf("Bucket ticked... current index: %v", jq.currentBucketIdx)
+		jq.mu.Lock()
+		if !jq.lastTick.Equal(time.Time{}) && t.After(jq.lastTick.Add(2*time.Second)) {
```

Comment:
I might be missing something obvious bu shouldn't this be `jq.lastTick.Add(1*time.Second)` ?

Reply:
The first problem is ticks aren't 100% precise. So when I tried this with 1 second, there were cases when we had slept just slightly more than a second. Prompting an incorrect log message, we were not running behind.

Thus, I considered it a missed scheduling slot if |t1| never fired, which would mean approximately two seconds between t0 and t2.
 
```
....----|t0|-----SCHED----|t1|-----SCHED-----|t2|----.... 
```

Another option might be to try to play the game of estimating just how precise those tickers can get 1ms, 10us, etc... and use that as the threshold. I didn't really want to play that fine-tuning game hence the 2sec (because that implied you had clearly missed your bucketing schedule slot).